### ISSUE TITLE
[ssr] "case" delta expanding fixpoints when simplification is needed since coq-8.6

### ISSUE THREAD

**Issue by chdoc** [coq/coq#12249]
#### Description of the problem

When applying the case tactic to a recursively defined proposition that needs to be reduced in order to exhibit a constructor of some inductive type, the fixpoint is delta-expanded (i.e., the full definition is shown.) 

```coq

From mathcomp Require Import all_ssreflect.
Set Implicit Arguments.
Unset Printing Implicit Defensive.
Import Prenex Implicits.

Inductive term := TT | And of term & term.
Fixpoint interp s := match s with TT => True | And s t => interp s /\ interp t end.
Definition foo s := And s s.

Goal forall s, interp (foo s) -> True.
  move => s. case.
  (* coq-8.5pl3/mathcomp-1.6.1 :  interp s -> interp s -> True *)

  (* coq-8.6/mathcomp-1.6.1 and the current development branches of coq/mathcomp *)
  (* 
  (fix interp (s0 : term) : Prop := match s0 with
                                    | TT => True
                                    | And s1 t => interp s1 /\ interp t
                                    end) s ->
  (fix interp (s0 : term) : Prop := match s0 with
                                    | TT => True
                                    | And s1 t => interp s1 /\ interp t
                                    end) s -> True
  *)
Abort.

(* Destruct does not unfold interp (neither in 8.5pl3 nor in 8.6) *)
Goal forall s, interp (foo s) -> True.
  move => s. destruct 1 as [A B].
Abort.
```
The new behaviour leads to a lot less readable goal states and causes subsequent rewrites to fail. The problem does not occur when using "simpl" before "case".

I reported this on the mathcomp issue tracker a long time ago, but then neglected to transfer the [issue](https://github.com/math-comp/math-comp/issues/101) to the Coq bug tracker. The issue persists.

#### Coq Version

The Coq Proof Assistant, version 8.12+alpha (May 2020)
compiled on May 4 2020 12:21:30 with OCaml 4.09.0

**Comment by chdoc** [comment]
@gares , This has been around forever. Is this an issue of the `case` tactic in ssreflect, or something else?

**Comment by gares** [comment]
I've no memory of having changed the code that "reduces an hyp to an inductive" that is called by case. At some point some of these APIs were hijacked by the refolding reduction machine with performance hits everywhere, and then this hijacking was undone here and there. This could explain the change in behavior.

I concur with you that case could use a less efficient but more user friendly reduction machinery. But that refolding machine I was talking about almost never refolds, it is, frankly, broken by design. It is the machine behind the cbn thing, it refolds only in simpler cases that what simpl does, but sometimes refolds even things that it never unfolded... 

So I'm a bit unsure unsure, the "easy fix" seems worse than the actual problem to me.

**Comment by chdoc** [comment]
I agree that this is not the most pressing issue, not just because there is an easy workaround (i.e, inserting a  `/=` before the case analysis). On the other hand, `destruct` refolds the fixpoint as expected. Not sure what's the right thing to do here.
